crates/turbo-tasks-memory/src/edges_set.rs

```diff
@@ -0,0 +1,592 @@
+        // We verified that the entry is present, so any non-complex case is easier to
+        // handle
+        match entry {
+            EdgeEntry::Output => match self {
+                EdgesEntry::Output => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::ChildAndOutput => {
+                    *self = EdgesEntry::Child;
+                    return true;
+                }
+                EdgesEntry::OutputAndCell0(type_id) => {
+                    *self = EdgesEntry::Cell0(*type_id);
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Child => match self {
+                EdgesEntry::Child => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::ChildAndOutput => {
+                    *self = EdgesEntry::Output;
+                    return true;
+                }
+                EdgesEntry::ChildAndCell0(type_id) => {
+                    *self = EdgesEntry::Cell0(*type_id);
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Cell(_) => match self {
+                EdgesEntry::Cell0(_) => {
+                    *self = EdgesEntry::Empty;
+                    return true;
+                }
+                EdgesEntry::OutputAndCell0(_) => {
+                    *self = EdgesEntry::Output;
+                    return true;
+                }
+                EdgesEntry::ChildAndCell0(_) => {
+                    *self = EdgesEntry::Child;
+                    return true;
+                }
+                _ => {}
+            },
+            EdgeEntry::Collectibles(_) => {}
+        }
+        if let EdgesEntry::Complex(set) = self {
+            if set.remove(&entry) {
+                self.simplify();
+                return true;
+            }
+        }
+        false
+    }
+
+    fn shrink_to_fit(&mut self) {
+        if let EdgesEntry::Complex(set) = self {
+            set.shrink_to_fit();
+        }
+    }
+
+    fn simplify(&mut self) {
+        if let EdgesEntry::Complex(set) = self {
+            match set.len() {
+                0 => {
+                    *self = EdgesEntry::Empty;
+                }
+                1 => {
+                    let entry = set.iter().next().unwrap();
+                    if matches!(
+                        entry,
+                        EdgeEntry::Output
+                            | EdgeEntry::Child
+                            | EdgeEntry::Cell(CellId { index: 0, .. })
+                    ) {
+                        *self = EdgesEntry::from(*entry);
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct TaskEdgesSet {
+    edges: Box<AutoMap<TaskId, EdgesEntry, BuildHasherDefault<FxHasher>>>,
```

Comment:
I don't think you should box this yourself. You should leave it up to the caller whether or not they want to box `TaskEdgesSet`.